dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UIBasicTest.java

```diff
@@ -0,0 +1,145 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package io.openliberty.microprofile.openapi.ui.internal.fat.tests;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.startsWith;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+import java.time.Duration;
+
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.openqa.selenium.By;
+import org.openqa.selenium.NoSuchElementException;
+import org.openqa.selenium.SearchContext;
+import org.openqa.selenium.TimeoutException;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeOptions;
+import org.openqa.selenium.remote.RemoteWebDriver;
+import org.openqa.selenium.support.ui.FluentWait;
+import org.testcontainers.Testcontainers;
+import org.testcontainers.containers.BrowserWebDriverContainer;
+import org.testcontainers.containers.BrowserWebDriverContainer.VncRecordingMode;
+import org.testcontainers.containers.VncRecordingContainer.VncRecordingFormat;
+
+import com.ibm.websphere.simplicity.ShrinkHelper;
+import com.ibm.websphere.simplicity.ShrinkHelper.DeployOptions;
+import com.ibm.ws.fat.util.Props;
+
+import componenttest.annotation.Server;
+import componenttest.containers.SimpleLogConsumer;
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.topology.impl.LibertyServer;
+import io.openliberty.microprofile.openapi.ui.internal.fat.app.TestResource;
+
+/**
+ * A basic test that we can open the UI and that it appears to have loaded correctly
+ */
+@RunWith(FATRunner.class)
+public class UIBasicTest {
+
+    /** Wait for "long" tasks like initial page load or making a test request to the server */
+    private static final Duration LONG_WAIT = Duration.ofSeconds(30);
+
+    /** Wait for "short" tasks which shouldn't need to wait for anything, like selecting a drop down */
+    private static final Duration SHORT_WAIT = Duration.ofSeconds(5);
+
+    public static final String APP_NAME = "app";
+
+    @Server("openapi-ui-test")
+    public static LibertyServer server;
+
+    @Rule
```

Comment:
I did initially have it as a `@ClassRule`, but `@Rule` seems to be the recommended way to do it for the web driver containers. If I use `@ClassRule`, the option to save a recording only for failing tests does not work.

Unfortunately the recording is pretty essential for working out what went wrong when a test fails. Maybe I should try adding a second test to check what the performance is like. If it's bad it might be preferable to always save the recording.

Reply:
For our tests, we don't do a recording - we take screenshots at critical points and on failure to use for debug.  (TakesScreenshot) Our tests are much larger and longer so a recording would be enormous. Curious where you saw that the @Rule is the recommended way? We are using classrule in most places.

Reply:
[Selenium recommends a fresh browser per test](https://www.selenium.dev/documentation/test_practices/encouraged/fresh_browser_per_test/)

All of the [testcontainers examples use `@Rule`](https://www.testcontainers.org/modules/webdriver_containers/) and it claims that creation of browser containers is "fast".

I am still a little concerned about the performance of using `@Rule` with our infrastructure. I'll be sure to keep an eye on it when I add more tests.

Taking a screenshot does seem more practical than recording every test, I will also look into that if we have performance or space issues. The recording was surprisingly small but I still wouldn't want to record passing tests.